File: Leetcode/757. Set Intersection Size At Least Two.cpp
class Solution {
public:
    static bool comp(vector<int>&a, vector<int>&b){
        if(a[1]==b[1]){
            return a[0]>b[0];
        }
        return a[1]<b[1];
    }
    int intersectionSizeTwo(vector<vector<int>>& intervals) {
        sort(intervals.begin(),intervals.end(),comp);
        int Last=-1,secondLast=-1,ans=0;
        for(auto i:intervals){
            if(i[0]<=secondLast){
                //all two are included in the interval so no problem.
                //nothing to do
                continue;
            }else if(i[0]>secondLast && i[0]<=Last){
                //one included need another one 
                secondLast=Last;
                Last=i[1];
                ans++;
            }else{
                //no one is included two numbers have to be included
                ans+=2;
                Last=i[1];
                secondLast=i[1]-1;
            }
        }
        return ans;
    }
};

File: Leetcode/757. Set Intersection Size At Least Two.py
class Solution:
    def intersectionSizeTwo(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x:(x[1],-x[0]))
        secondLast,Last,ans=-1,-1,0
        for (a,b) in intervals:
            if a <= secondLast:
                #already 2 are included
                #nothing to do 
                pass
            elif a>secondLast and a<=Last:
                #only one included we have to increase ans by 1
                ans+=1
                secondLast=Last
                Last=b
            else:
                #no interval included
                ans+=2
                secondLast=b-1
                Last=b
        return ans
        
